### PR TITLE
Adding AVX2 to significantly improve speed on arrays of Int32

### PR DESCRIPTION
**Pull request by HelifeWasTaken**

### CODE REVIEW

**Review by Matthieu-Sauer — CHANGES_REQUESTED** [review]
I will check this evening if functionnality is working as expected.
You should create a "simd.hpp" or "intrinsics.hpp" which will decide (compile time) which intrinsics to use for a better extendability (we might want to target machines which do not have avx2 in the lobg run).

Also, I did not ask, however was this part of the code critical enough for intrinsics to be used? last time I benchmarked perlin noise generation and chunk accesses were the maik bottlenecks (did not check nbt as server crashes when I join on my machine)

File: cubic-server/nbt.cpp [hunk 1]
```diff
@@ -373,13 +376,53 @@ std::shared_ptr<IntArray> nbt::parseIntArray(uint8_t *&at, const uint8_t *end, b
...
     }
+#else
+    // The SIMD implementation should be at least 8x times faster
+    // When arrays are of at least size >= 4
+
+    data.resize(size);
+    int32_t unaligned;
+    int32_t i = 0;
+
+    if (size >= 8) goto aligned8;
```
[the remark this answers is not in the document]
Comment (Matthieu-Sauer) [reply cubic-server/nbt.cpp:396]: I would even suggest templated ones for same reason, easier to extend :)

File: cubic-server/nbt.cpp [hunk 2]
```diff
@@ -1,6 +1,10 @@
 #include "nbt.hpp"
 #include <optional>
 
+#ifdef __AVX2__
```
[the remark this answers is not in the document]
Comment (HelifeWasTaken) [reply cubic-server/nbt.cpp:4]: > I will check this evening if functionnality is working as expected. You should create a "simd.hpp" or "intrinsics.hpp" which will decide (compile time) which intrinsics to use for a better extendability (we might want to target machines which do not have avx2 in the lobg run).
> 
> Also, I did not ask, however was this part of the code critical enough for intrinsics to be used? last time I benchmarked perlin noise generation and chunk accesses were the maik bottlenecks (did not check nbt as server crashes when I join on my machine)

I actually tried to run the CubicMC server on the latest release (not the master tho) and found out that a lot of the time was used reading nbt list values so I guess -\_(' -')_/-

**Comment by huntears** [comment]
@HelifeWasTaken If you can prove there is a significant performance gain I would be glad to merge this, but if there is not it just looks like more maintenability problems.

**Comment by HelifeWasTaken** [comment]
> 
Might do it when I have time np

**Comment by huntears** [comment]
Closing this PR as we are currently moving away from our homemade NBT library to a properly coded one.

**Pull request closed without merging**